Feature/files list occ command

yemkareems opened the pull request:
files:list implemented based on files:scan
List of possible usage

all user based search
```occ files:list --all```

path based search
```occ files:list --path /admin/files```
scans the directory and sub directory also

Filter by type
```
occ files:list --path /admin/files/Media --type video
occ files:list --path /admin/files/Media --type image
occ files:list --path /admin/files/ --type application
```

Filter by file size
```
occ files:list --path /admin/files/Media --minSize 9622690
occ files:list --path /admin/files/Media --maxSize 9622690
```

default sort order is desc
```occ files:list --all --sort name```
to do asc
```occ files:list --all --sort name --order ASC```

Non recursive scan and list
```occ files:list --path /admin/files/ --shallow```
The above does not scan sub directories recursively

All the above params can be combined to have different result sets


* Resolves: # <!-- related github issue -->

## Summary


## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

yemkareems committed:
files:list added as a occ command. filter added based on type and size for files. path or --all is required to search files in the path or based on all users

yemkareems committed:
basic sorting added to list, could be improvised

yemkareems committed:
listing files and folders separately as sorting is not working as expected

yemkareems committed:
general refactoring to remove unused variables and functions, edge case in file info if no files in dir covered

yemkareems commented:
@AndyScherzinger @sorbaugh Please find the above file list implemented based on files:scan. I am not able add you as reviewers so tagged both of you in the comment here.

yemkareems committed:
Update ListFiles.php

Psalm errors fixed

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
composer run cs:fix

composer run cs:fix

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
Update ListFiles.php

exectime variable removed to fix psalm error

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
type hinting

type hinting corrected

Signed-off-by: yemkareems <yemkareems@gmail.com>

artonge commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				$this->fileInfo[] = $this->getNodeInfo($node, $path);
+			}
+			$this->abortIfInterrupted();
+			if ($scanMetadata !== null) {
+				$this->filesMetadataManager->refreshMetadata(
+					$node,
+					($scanMetadata !== '') ? IFilesMetadataManager::PROCESS_NAMED : IFilesMetadataManager::PROCESS_LIVE | IFilesMetadataManager::PROCESS_BACKGROUND,
+					$scanMetadata
+				);
+			}
```
Not needed if the command is only about listing files.

yemkareems replied:
Removed the same

artonge commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+			if ($backgroundScan) {
+				$scanner->backgroundScan($path);
+			} else {
+				$scanner->scan($path, $recursive, $homeOnly ? [$this, 'filterHomeMount'] : null);
```
This is probably too heavy for what you are trying to do. The scan method will sync the file system state to the database. Here, we probably only want to list the files.

```php
$userFolder = $this->rootFolder->getUserFolder($user);
$nodes = $userFolder->getDirectoryListing();
```

And then you can list the files recursively.

yemkareems replied:
Replaced the scan method with the suggested getUserFolder

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
+ *
+ * @author Bart Visscher <bartv@thisnet.nl>
+ * @author Blaok <i@blaok.me>
+ * @author Christoph Wurst <christoph@winzerhof-wurst.at>
+ * @author Daniel Kesselberg <mail@danielkesselberg.de>
+ * @author J0WI <J0WI@users.noreply.github.com>
+ * @author Joas Schilling <coding@schilljs.com>
+ * @author Joel S <joel.devbox@protonmail.com>
+ * @author Jörn Friedrich Dreyer <jfd@butonic.de>
+ * @author martin.mattel@diemattels.at <martin.mattel@diemattels.at>
+ * @author Maxence Lange <maxence@artificial-owl.com>
+ * @author Robin Appelman <robin@icewind.nl>
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
+ * @author Thomas Müller <thomas.mueller@tmit.eu>
+ * @author Vincent Petry <vincent@nextcloud.com>
```
Hi @artonge , Thank you for the review. My local development VM is not allowing me to load php strom and edit files. I will set it up over the weekend and resolve most of the PR comments and I will let you know once I am done. Sorry about the delay. Thank you once again

yemkareems replied:
Hi @artonge, Though i had issues with php strom i have made most of the changes and run the command and tested it for scenarios and also did a composer run cs:fix since i used gedit to edit the file.

artonge replied:
The copy noticeright still lists unrelated authors.

yemkareems replied:
@artonge Removed the author's list

yemkareems committed:
PR review comments by Louis done

Scan files replaced with listFiles and all methods refactored accordingly

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				InputOption::VALUE_OPTIONAL,
+				'Generate metadata for all scanned files; if specified only generate for named value',
+				''
+			)
```
Removed the generate-metadata

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				'all',
+				null,
+				InputOption::VALUE_NONE,
+				'will rescan all files of all known users'
```
Replaced scan with list across the board

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				'unscanned',
+				null,
+				InputOption::VALUE_NONE,
+				'only scan files which are marked as not fully scanned'
```
Removed the same

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				'home-only',
+				null,
+				InputOption::VALUE_NONE,
+				'only scan the home storage, ignoring any mounted external storage or share'
```
Replaced scan with list

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				'path',
+				'p',
+				InputArgument::OPTIONAL,
+				'limit rescan to this path, eg. --path="/alice/files/Music", the user_id is determined by the path and the user_id parameter and --all are ignored'
```
--all is removed and path is left as it is

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				'all',
+				null,
+				InputOption::VALUE_NONE,
+				'will rescan all files of all known users'
```
Removed the all param

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,406 @@
+				'shallow',
+				null,
+				InputOption::VALUE_NONE,
+				'do not scan folders recursively'
```
Shallow is removed as scan is no longer used

icewind1991 commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,334 @@
+
+		$this->setName("files:list")
+			->setDescription("list filesystem")
+			->addOption(
+				"path",
+				"p",
+				InputArgument::OPTIONAL,
+				'limit list to this path, eg. --path="/alice/files/Music", the user_id is determined by the path and the user_id parameter'
+			)
```
it makes more sense to have this as an argument instead of an option since it's always needed

yemkareems replied:
Hi @icewind1991 Path is made an argument and user logic changed accordingly

yemkareems committed:
path made a argument and user extraction logic based on path changed

path made a argument and user extraction logic based on path changed

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
help doc content changed to start with all first caps

help doc content changed to start with all first caps

Signed-off-by: yemkareems <yemkareems@gmail.com>

artonge commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,334 @@
+		}
+
+		# check quantity of users to be process and show it on the command line
+		$users_total = count($users);
```
There will always be only one user now that the `--all` flag was removed.

yemkareems replied:
@artonge Changed the users logic to have only one user or user not found error in case of unknown user

yemkareems committed:
removed the user loop since only one user is there and copy right authors removed

removed the user loop since only one user is there and copy right authors removed

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
type hinting corrected

type hinting corrected

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
cs fix ran 

cs fix ran 

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
array multisort assigning to var and passing it

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
userFolder type hint to fix psalm error

userFolder type hint to fix psalm error

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,334 @@
+	): void {
+		try {
+			$userFolder = $this->rootFolder->get($path);
+			$files = $userFolder->getDirectoryListing();
```
I could not get the psalm going in my local dev env but did fix few of the issues in the latest commit

AndyScherzinger commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,304 @@
+<?php
+/**
+ * @copyright Copyright (c) 2024, ownCloud, Inc.
```
```suggestion
 * @copyright Copyright (c) 2024, Nextcloud GmbH
```

yemkareems committed:
Update apps/files/lib/Command/ListFiles.php

Co-authored-by: Andy Scherzinger <info@andy-scherzinger.de>
Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
updated the description of path

updated the description of path

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,304 @@
+			->addArgument(
+				"path",
+				InputArgument::REQUIRED,
+				'Limit list to this path, eg. path="/alice/files/Music", the user_id is determined by the path parameter'
```
@icewind1991 Updated the description of the path argument

icewind1991 replied:
I mean the `path=...` doesn't match anymore as it's no longer an option, so you no longer specify the `path=` part

yemkareems replied:
Hi @icewind1991 not sure if i understood it correctly because i still do occ files:list path=/admin/files/Media --type=video where path is required and the above lists all the video files in the path

icewind1991 commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,304 @@
+			"perm" => $node->getPermissions(),
+			"owner" => $node->getOwner()->getDisplayName(),
+			"created-at" => $node->getCreationTime(),
+			"type" => $node->getMimePart(),
```
would be good to have some logic to show `directory` here instead of the current `httpd` (the mimetype of directories is `httpd/unix-directory` in nc)

yemkareems replied:
Hi @icewind1991 made the change to have a check for 
		if($node->getMimetype() == FileInfo::MIMETYPE_FOLDER) {
			$nodeInfo['type'] = 'directory';
		}

yemkareems committed:
making changes to list folders as directory type

making changes to list folders as directory type

Signed-off-by: yemkareems <yemkareems@gmail.com>

yemkareems committed:
cs fix run for the file

cs fix run for the file

Signed-off-by: yemkareems <yemkareems@gmail.com>

artonge commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,311 @@
+		?int $maxSize = 0
+	): void {
+		try {
+			/** @ var OC\Files\Node\Folder $userFolder **/
```
Psalm is complaining because `$this->rootFolder->get` return type is `Node`, so we need to tell psalm that we know what we are doing, and the `$userFolder` is a `Folder`. I think the following should work:

```suggestion
			/** @var OCP\Files\Folder $userFolder **/
```

That, or add the following check:

```php
if (!($userFolder instanceof Folder)) {
    // throw or return.
}
```

yemkareems committed:
telling psalm correct path of folder

artonge commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,311 @@
+		?int $maxSize = 0
+	): void {
+		try {
+			/** @ var OCP\Files\Folder $userFolder **/
```
You need to remove the extra space between `@` and `var`

```suggestion
			/** @var OCP\Files\Folder $userFolder **/
```

yemkareems committed:
feat: tell psalm the correct type associated with userFolder

yemkareems committed:
feat: tell psalm the correct type associated with userFolder

susnux commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,311 @@
+ * @copyright Copyright (c) 2024, Nextcloud GmbH
+ *
+ * @author Kareem <yemkareems@gmail.com>
+ * @license AGPL-3.0
```
We nowadays use `AGPL-3.0-or-later` if you ok with it:

```suggestion
 * @license AGPL-3.0-or-later
```

susnux commented on apps/files/lib/Command/ListFiles.php:
```diff
@@ -0,0 +1,311 @@
+ * @copyright Copyright (c) 2024, Nextcloud GmbH
+ *
+ * @author Kareem <yemkareems@gmail.com>
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
+ */
```
We nowadays use `AGPL-3.0-or-later`, if you ok with it:

```suggestion
  * @license AGPL-3.0-or-later
 *
 * This code is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License, version 3,
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 *
 */
```

yemkareems replied:
Hi @susnux Changed the wording as suggested

yemkareems committed:
feat: license wording changed as suggested

artonge commented:
Closing as it move here: https://github.com/nextcloud/server/pull/46352
